### DISCUSSION TITLE
PFC for pb2019 R3

### DISCUSSION THREAD

**Discussion by mjl38** [OpenSourcePFCLibraries/2019#20] [Q&A] [answered]
Hi,
I'm not aware the PFC have been adapted for pb2019 R**3**.
I don't think a lot would have to be changed, but I might be wrong.
Can anyone confirm this?
regards
MiguelL

**Comment by brucearmstrong2** [comment]
There is nothing I'm aware of in R3 that would require any modifications from earlier 2019 versions.  I'm using R3 and it's working fine for me.

What is your starting point?  2019 or something earlier.

**Reply by mjl38** [reply]
Hi Bruce, I'm not really sure if I'm using an adapted version of 2017 or an earlier version of 2019.
Tomorrow, once this horrible tooth pain is over, I'll have a look.

**Reply by mjl38** [reply]
This is one of the things that should change. I did a change, but cannot seem to find it. See the other discussion I've created.
pfc_n_cst_platform.constructor event:
......
ls_PBVMName = this.of_getPBVMName()
choose case ls_PBVMName
	case "pbvm.dll"  // pb2019 R3 (and most probably later versions too)
		is_ClassName[1] = "FNWND3" // mdi window
		is_ClassName[2] = "FNWNS3" // response window
	case "pbvm170.dll" // pb2017
		is_ClassName[1] = "FNWND3170" // mdi window
		is_ClassName[2] = "FNWNS3170" // response window
		is_ClassName[3] = "AfxMDIFrame100su" // pb2017 for "dockable MDI" 
	case "pbvm190.dll"  // pb2019 R1 / R2
		is_ClassName[1] = "FNWND3190" // mdi window
		is_ClassName[2] = "FNWNS3190" // response window
		// dockable MDI is now obsolete in pb2019 R2, commented:
		//	is_ClassName[3] = "AfxMDIFrame100su" // pb2019 for "dockable MDI", same as PB2017
	case "pbvm126" // pb 12.6
		is_ClassName[1] = "FNWND3126" // mdi window
		is_ClassName[2] = "FNWNS3126" // response window
	case "pbvm125" // pb 12.5
		is_ClassName[1] = "FNWND3125" // mdi window
		is_ClassName[2] = "FNWNS3125" // response window	
end choose
....
regards

**Reply by mjl38** [reply]
So, to resume my previous post:
2019 R3 should use:
is_ClassName[1] = "FNWND3" // mdi window
is_ClassName[2] = "FNWNS3" // response window

and previous versions of 2019 should use:
is_ClassName[1] = "FNWND3190" // mdi window
is_ClassName[2] = "FNWNS3190" // response window

**Reply by mjl38** [reply]
Found the changes in ws_objects.
Shouldn't they be in the ZIP too?
regards

**Reply by mjl38** [reply]
So that's really funny. I'm asking if the libraries have been adapted for R3 and as a result I did that myself... Getting old.

**Comment by bruce-armstrong** [comment, marked as answer]
I just created a new release that contains a new ZIP file with PBLs that should contain all the latest changes.

**Reply by mjl38** [reply]
That´s great news. Thank you Bruce!

**Reply by jojo5791** [reply]
Hi Bruce, I am a newbie here. May I know if i want to change codes inside pfc_n_cst_error (of_messages), can i just change/update a pfc_n_cst_error - of_message?or what should i do? Thank you. Hope to hear from you soon.

**Reply by mjl38** [reply]
In the readme it says that if you want to contribute, you can watch this video [https://vimeo.com/478243534](https://vimeo.com/478243534)

**Reply by jojo5791** [reply]
noted. But, what i changed is customized which i need to change it in order to adapt the language localization to take effect. Therefore, i dnt think this contribution/changes is appropriate to submit here. My question is, what should i do if i make changes to pfc_n_cst_error and in future, any new set of PFC may overwrite it, how to avoid this?  Hope i didnt confused you and hope to hear from you soon. Thank you very much in advance.

**Reply by mjl38** [reply]
Aha, in that case you have to do your changes in the pfE libraries, not the pfC. When there's a new version of pfc classes, you keep your modified pfE libraries and hook those up with the newer version of pfE classes.
So in **n_cst_error** you would override the ancestor class of the **pfc_n_cst_error** ancestor. If it's a function you just copy the code of the pfc object and paste it in the pfe object, and do your changes. If it's an event that you have to override, you do the same thing, but don't forget to uncheck "extend ancestor class" when doing a rightmousclick somewhere in the code. Take notice that anytime that you delete all the code in an overriden event, that property is automatically checked again.
regards

**Reply by mjl38** [reply]
Also, you might have more success with questions like this on Appeon's Q&A site. There's regretfully not many people contributing here.

**Reply by jojo5791** [reply]
Hi Miguel,  Thanks a lot for your prompt feedback. It is very helpful. I appreciate it very much. I have one more question, if I override the ancestor class of the pfc_n_cst_error, at pfe object, in future, with new PFC, will it overwrite what i override at n_cst_error? Because i noticed, new PFC sources/libraries list, always provide pfcapsrv and pfeapsrv.

**Reply by mjl38** [reply]
Hi Jane,
Yw. If you overwrite your existing 'old' pfe pbl files, you'll have lost your changes. The pfe classes that come with any new PFC version, won't / shouldn't have any code in them. Sometimes though, your old pfe classes will be missing some new object if one has been created in the PFC or there can be differences in function / event parameters, new events and functions, etc. So my recommendation is as follows: Make sure that all of your changes in pfe classes have some kind of a special comment. That way you can hookup both your old and new pfe libraries in an application and then copy your changes in the old pfe to the new pfe. Being able to search on that special comment, will make things a lot easier. When finished, you simply remove your old pfe classes from the library list.
regards

**Reply by jojo5791** [reply]
Noted with many thanks Miguel.  You are my angel. :)

**Reply by mjl38** [reply]
Lol, you make me blush! Anyway, please use the [https://community.appeon.com/index.php/qna/q-a](https://community.appeon.com/index.php/qna/q-a), you'll get a lot of support from several people.
Take care,
MiguelL

**Reply by jojo5791** [reply]
Hi Miguel, noted. haha.... i have an idea, can i create another new library, named as pfdapsrv, and inherit the n_cst_error from pfeapsrv as pfd_n_cst_error. Then do the rest as you mentioned above. Therefore, in future, any new/overwrite of pfcapsrv and pfeapsrv, would not overwrite my customized changes? Is that correct? Thanks,

**Reply by mjl38** [reply]
Yes, that's correct, but not really what I'd recommend doing. It adds another inheritance layer to the already bloated pfc classes. It's what a some companies do though.

**Reply by jojo5791** [reply]
noted. tqvm,

**Comment by start2ar** [comment]
thanks!

Il 08/04/2021 18.42, Bruce Armstrong ha scritto:
>
> I just created a new release that contains a new ZIP file with PBLs 
> that should contain all the latest changes.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/OpenSourcePFCLibraries/2019/discussions/20#discussioncomment-585810>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AOG2AWMDC36TZPU4VKU7UNLTHXMIZANCNFSM42QRCCHQ>.
>
-- 
<http://www.start2.it>
*Tel. 0575984910* 	
*Via F.lli Lumiere, 19* 	***@***.*** ***@***.***>
*52100 AREZZO (AR)* 	www.start2.it <http://www.start2.it>
